File: Java/_133_Clone_Graph/Solution.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public static Map<UndirectedGraphNode, UndirectedGraphNode> hash = new HashMap<UndirectedGraphNode, UndirectedGraphNode>();
    
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null)
            return null;
        if (hash.containsKey(node))
            return hash.get(node);
        hash.put(node, new UndirectedGraphNode(node.label));
        for (UndirectedGraphNode val : node.neighbors) {
            UndirectedGraphNode temp = hash.get(node);
            temp.neighbors.add(cloneGraph(val));
            hash.put(node, temp);
        }
        return hash.get(node);
    }
}

File: Python/_133_Clone_Graph/solution.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def __init__(self):
        self.map = {}
    
    def cloneGraph(self, node):
        if node == None:
            return None
        if self.map.has_key(node):
            return self.map[node]
        self.map[node] = UndirectedGraphNode(node.label)
        for val in node.neighbors:
            self.map[node].neighbors.append(self.cloneGraph(val))
        return self.map[node]
